File: Tests/Tests/STPSetupIntentTest.m
//
//  STPSetupIntentTest.m
//  StripeiOS Tests
//
//  Created by Yuki Tokuhiro on 6/27/19.
//  Copyright © 2019 Stripe, Inc. All rights reserved.
//

#import <XCTest/XCTest.h>

#import "STPFixtures.h"
#import "STPTestUtils.h"

@interface STPSetupIntentTest : XCTestCase

@end

@implementation STPSetupIntentTest

#pragma mark - Description Tests

- (void)testDescription {
    STPSetupIntent *setupIntent = [STPFixtures setupIntent];
    
    XCTAssertNotNil(setupIntent);
    NSString *desc = setupIntent.description;
    XCTAssertTrue([desc containsString:NSStringFromClass([setupIntent class])]);
    XCTAssertGreaterThan(desc.length, 500UL, @"Custom description should be long");
}

#pragma mark - STPAPIResponseDecodable Tests

- (void)testDecodedObjectFromAPIResponseRequiredFields {
    NSDictionary *fullJson = [STPTestUtils jsonNamed:STPTestJSONSetupIntent];
    
    XCTAssertNotNil([STPSetupIntent decodedObjectFromAPIResponse:fullJson], @"can decode with full json");
    
    NSArray<NSString *> *requiredFields = @[
                                            @"id",
                                            @"client_secret",
                                            @"livemode",
                                            @"status",
                                            ];
    
    for (NSString *field in requiredFields) {
        NSMutableDictionary *partialJson = [fullJson mutableCopy];
        
        XCTAssertNotNil(partialJson[field], @"json should contain %@", field);
        [partialJson removeObjectForKey:field];
        
        XCTAssertNil([STPSetupIntent decodedObjectFromAPIResponse:partialJson], @"should not decode without %@", field);
    }
}

- (void)testDecodedObjectFromAPIResponseMapping {
    NSDictionary *response = [STPTestUtils jsonNamed:@"SetupIntent"];
    STPSetupIntent *setupIntent = [STPSetupIntent decodedObjectFromAPIResponse:response];
    
    XCTAssertEqualObjects(setupIntent.stripeID, @"seti_123456789");
    XCTAssertEqualObjects(setupIntent.clientSecret, @"seti_123456789_secret_123456789");
    XCTAssertEqualObjects(setupIntent.created, [NSDate dateWithTimeIntervalSince1970:123456789]);
    XCTAssertEqualObjects(setupIntent.customerID, @"cus_123456");
    XCTAssertEqualObjects(setupIntent.paymentMethodID, @"pm_123456");
    XCTAssertEqualObjects(setupIntent.stripeDescription, @"My Sample SetupIntent");
    XCTAssertFalse(setupIntent.livemode);
    // nextAction
    XCTAssertNotNil(setupIntent.nextAction);
    XCTAssertEqual(setupIntent.nextAction.type, STPIntentActionTypeRedirectToURL);
    XCTAssertNotNil(setupIntent.nextAction.redirectToURL);
    XCTAssertNotNil(setupIntent.nextAction.redirectToURL.url);
    NSURL *returnURL = setupIntent.nextAction.redirectToURL.returnURL;
    XCTAssertNotNil(returnURL);
    XCTAssertEqualObjects(returnURL, [NSURL URLWithString:@"payments-example://stripe-redirect"]);
    NSURL *url = setupIntent.nextAction.redirectToURL.url;
    XCTAssertNotNil(url);
    
    XCTAssertEqualObjects(url, [NSURL URLWithString:@"https://hooks.stripe.com/redirect/authenticate/src_1Cl1AeIl4IdHmuTb1L7x083A?client_secret=src_client_secret_DBNwUe9qHteqJ8qQBwNWiigk"]);
    XCTAssertEqualObjects(setupIntent.paymentMethodID, @"pm_123456");
    XCTAssertEqual(setupIntent.status, STPSetupIntentStatusRequiresAction);
    XCTAssertEqual(setupIntent.usage, STPSetupIntentUsageOffSession);
    
    XCTAssertEqualObjects(setupIntent.paymentMethodTypes, @[@(STPPaymentMethodTypeCard)]);
    
    // lastSetupError
    
    XCTAssertNotNil(setupIntent.lastSetupError);
    XCTAssertEqualObjects(setupIntent.lastSetupError.code, @"setup_intent_authentication_failure");
    XCTAssertEqualObjects(setupIntent.lastSetupError.docURL, @"https://stripe.com/docs/error-codes/setup-intent-authentication-failure");
    XCTAssertEqualObjects(setupIntent.lastSetupError.message, @"The latest attempt to set up the payment method has failed because authentication failed.");
    XCTAssertNotNil(setupIntent.lastSetupError.paymentMethod);
    XCTAssertEqual(setupIntent.lastSetupError.type, STPSetupIntentLastSetupErrorTypeInvalidRequest);
    
    XCTAssertNotEqual(setupIntent.allResponseFields, response, @"should have own copy of fields");
}


@end

File: Stripe/StripeiOSTests/STPSetupIntentTest.swift
//  Converted to Swift 5.8.1 by Swiftify v5.8.28463 - https://swiftify.com/
//
//  STPSetupIntentTest.m
//  StripeiOS Tests
//
//  Created by Yuki Tokuhiro on 6/27/19.
//  Copyright © 2019 Stripe, Inc. All rights reserved.
//
@_spi(STP) @testable import StripePayments

class STPSetupIntentTest: XCTestCase {
    // MARK: - Description Tests

    func testDescription() {
        let setupIntent = STPFixtures.setupIntent()

        XCTAssertNotNil(setupIntent)
        let desc = setupIntent.description
        XCTAssertTrue(desc.contains(NSStringFromClass(type(of: setupIntent).self)))
        XCTAssertGreaterThan(desc.count, 500, "Custom description should be long")
    }

    // MARK: - STPAPIResponseDecodable Tests

    func testDecodedObjectFromAPIResponseRequiredFields() {
        let fullJson = STPTestUtils.jsonNamed(STPTestJSONSetupIntent)

        XCTAssertNotNil(STPSetupIntent.decodedObject(fromAPIResponse: fullJson), "can decode with full json")

        let requiredFields = [
            "id",
            "client_secret",
            "livemode",
            "status",
        ]

        for field in requiredFields {
            var partialJson = fullJson! as [AnyHashable: Any]

            XCTAssertNotNil(partialJson[field])
            partialJson.removeValue(forKey: field)

            XCTAssertNil(STPSetupIntent.decodedObject(fromAPIResponse: partialJson))
        }
    }

    func testDecodedObjectFromAPIResponseMapping() {
        let setupIntentJson = STPTestUtils.jsonNamed("SetupIntent")!
        guard let setupIntent = STPSetupIntent.decodedObject(fromAPIResponse: setupIntentJson) else { XCTFail(); return }

        XCTAssertEqual(setupIntent.stripeID, "seti_123456789")
        XCTAssertEqual(setupIntent.clientSecret, "seti_123456789_secret_123456789")
        XCTAssertEqual(setupIntent.created, Date(timeIntervalSince1970: 123456789))
        XCTAssertEqual(setupIntent.customerID, "cus_123456")
        XCTAssertEqual(setupIntent.paymentMethodID, "pm_123456")
        XCTAssertEqual(setupIntent.stripeDescription, "My Sample SetupIntent")
        XCTAssertFalse(setupIntent.livemode)
        // nextAction
        XCTAssertNotNil(setupIntent.nextAction)
        XCTAssertEqual(setupIntent.nextAction?.type, STPIntentActionType.redirectToURL)
        XCTAssertNotNil(setupIntent.nextAction?.redirectToURL)
        XCTAssertNotNil(setupIntent.nextAction?.redirectToURL?.url)
        let returnURL = setupIntent.nextAction?.redirectToURL?.returnURL
        XCTAssertNotNil(returnURL)
        XCTAssertEqual(returnURL, URL(string: "payments-example://stripe-redirect"))
        let url = setupIntent.nextAction?.redirectToURL?.url
        XCTAssertNotNil(url)

        XCTAssertEqual(url, URL(string: "https://hooks.stripe.com/redirect/authenticate/src_1Cl1AeIl4IdHmuTb1L7x083A?client_secret=src_client_secret_DBNwUe9qHteqJ8qQBwNWiigk"))
        XCTAssertEqual(setupIntent.paymentMethodID, "pm_123456")
        XCTAssertEqual(setupIntent.status, STPSetupIntentStatus.requiresAction)
        XCTAssertEqual(setupIntent.usage, STPSetupIntentUsage.offSession)

        XCTAssertEqual(setupIntent.paymentMethodTypes, [NSNumber(value: STPPaymentMethodType.card.rawValue)])

        // lastSetupError

        XCTAssertNotNil(setupIntent.lastSetupError)
        XCTAssertEqual(setupIntent.lastSetupError?.code, "setup_intent_authentication_failure")
        XCTAssertEqual(setupIntent.lastSetupError?.docURL, "https://stripe.com/docs/error-codes#setup-intent-authentication-failure")
        XCTAssertEqual(setupIntent.lastSetupError?.message, "The latest attempt to set up the payment method has failed because authentication failed.")
        XCTAssertNotNil(setupIntent.lastSetupError?.paymentMethod)
        XCTAssertEqual(setupIntent.lastSetupError?.type, STPSetupIntentLastSetupErrorType.invalidRequest)
    }

    // MARK: STPSetupIntentStatus extension tests

    func testStringFromStatus() {
        let expected: [STPSetupIntentStatus: String] = [
            .requiresPaymentMethod: "requires_payment_method",
            .requiresConfirmation: "requires_confirmation",
            .requiresAction: "requires_action",
            .processing: "processing",
            .succeeded: "succeeded",
            .canceled: "canceled",
            .unknown: "unknown",
        ]

        for (status, expectedString) in expected {
            let resultString = STPSetupIntentStatus.string(from: status)
            XCTAssertEqual(resultString, expectedString, "Expected \(status) to map to string '\(expectedString)', but got '\(resultString)'")
        }
    }
}
